### PR TITLE
[AST] PackExpansionMatcher: use common prefix/suffix algorithm for tuple matching

### PR DESCRIPTION
**Pull request by xedin**
- Common prefix/suffix should account for presence of labels
- Labeled parameters cannot appear in the region absorbed by
   a pack expansion type.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

File: include/swift/AST/PackExpansionMatcher.h [hunk 1]
```diff
@@ -63,49 +63,78 @@ class TuplePackMatcher {
...
 ///
 /// The invariant is that each list must only contain at most one pack
 /// expansion type. After collecting a common prefix and suffix, the
 /// pack expansion on either side asborbs the remaining elements on the
 /// other side.
-class ParamPackMatcher {
-  ArrayRef<AnyFunctionType::Param> lhsParams;
-  ArrayRef<AnyFunctionType::Param> rhsParams;
+class TypeListPackMatcher {
+  struct Element {
```
Comment (slavapestov) [inline include/swift/AST/PackExpansionMatcher.h:51]: Could we avoid conversion into the common Element representation by making the TypeListPackMatcher a template parameterized by the element type?

Comment (xedin) [reply include/swift/AST/PackExpansionMatcher.h:51]: Ah I think I know what you mean, we could parametrize over Element and then lazily convert while matching, I'll give it a try.

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**